cmd/libsnap-confine-private/string-utils.c

```diff
@@ -69,3 +70,29 @@ int sc_must_snprintf(char *str, size_t size, const char *format, ...)
 
 	return n;
 }
+
+void sc_string_append(char *buf, size_t buf_size, const char *str)
+{
+	// Set errno in case we die.
+	errno = 0;
+	if (buf == NULL) {
+		die("cannot append string: buffer is NULL");
+	}
+	if (str == NULL) {
+		die("cannot append string: string is NULL");
+	}
+	size_t buf_len = strlen(buf);
+	if (buf_len >= buf_size) {
+		// When writing tests for this I wrote a test that didn't initialize
+		// the buffer and was surprised to see the effect. While this test is
+		// not bulletproof it is a good sanity check to have. While the next
+		// test will always catch this issue the message is clearly distinct.
+		die("cannot append string: uninitialized buffer detected");
+	}
+	size_t str_len = strlen(str);
+	if (buf_len + str_len >= buf_size) {
+		die("cannot append string: buffer overflow of %td byte(s)",
+		    buf_len + str_len + 1 - buf_size);
+	}
+	memcpy(buf + buf_len, str, str_len + 1);
+}
```

Comment:
Done. I never thought about using strnlen and in retrospective it is much better and safer to do it this way.